custom-recipes/amazon-rekognition-object-detection-labeling/recipe.py

```diff
@@ -0,0 +1,123 @@
+input_folder = dataiku.Folder(input_folder_names[0])
+input_folder_is_s3 = input_folder.get_info().get("type", "") == "S3"
+if input_folder_is_s3:
+    logging.info("Input folder is on Amazon S3")
+    input_folder_access_info = input_folder.get_info().get("accessInfo", {})
+    input_folder_bucket = input_folder_access_info.get("bucket")
+    input_folder_root_path = str(input_folder_access_info.get("root", ""))[1:]
+
+output_dataset_names = get_output_names_for_role("output_dataset")  # mandatory output
+output_dataset = dataiku.Dataset(output_dataset_names[0])
+
+output_folder_names = get_output_names_for_role("output_folder")  # optional output
+output_folder = None
+if len(output_folder_names) != 0:
+    output_folder = dataiku.Folder(output_folder_names[0])
+
+recipe_config = get_recipe_config()
+api_configuration_preset = recipe_config.get("api_configuration_preset")
+api_quota_rate_limit = api_configuration_preset.get("api_quota_rate_limit")
+api_quota_period = api_configuration_preset.get("api_quota_period")
+parallel_workers = api_configuration_preset.get("parallel_workers")
+num_objects = int(recipe_config.get("num_objects", 1))
+if num_objects < 1:
+    raise ValueError("Number of objects must be greater than 1")
+minimum_score = int(recipe_config.get("minimum_score", 0) * 100)
+if minimum_score < 0 or minimum_score > 100:
+    raise ValueError("Minimum confidence score must be between 0 and 1")
+orientation_correction = bool(recipe_config.get("orientation_correction"))
+error_handling = ErrorHandlingEnum[recipe_config.get("error_handling")]
+
+client = get_client(api_configuration_preset)
+column_prefix = "object_api"
+
+
+# ==============================================================================
+# RUN
+# ==============================================================================
+
+image_path_list = [p for p in generate_path_list(input_folder) if supported_image_format(p)]
+input_df = pd.DataFrame(image_path_list, columns=[IMAGE_PATH_COLUMN])
+if len(input_df.index) == 0:
+    raise ValueError("No images of supported format (PNG or JPG) were found in the folder")
+
+
+@retry((RateLimitException, OSError), delay=api_quota_period, tries=5)
+@limits(calls=api_quota_rate_limit, period=api_quota_period)
+def call_api_object_detection(row: Dict, num_objects: int, minimum_score: int, orientation_correction: bool) -> AnyStr:
+    image_path = row.get(IMAGE_PATH_COLUMN)
+    pil_image = None
+    if input_folder_is_s3:
+        image_request = {"S3Object": {"Bucket": input_folder_bucket, "Name": input_folder_root_path + image_path}}
+    else:
+        with input_folder.get_download_stream(image_path) as stream:
+            image_request = {"Bytes": stream.read()}
+            pil_image = Image.open(BytesIO(image_request["Bytes"]))
+    if orientation_correction:
+        detected_orientation = client.recognize_celebrities(Image=image_request).get("OrientationCorrection", "")
```

Comment:
The `api_parallelizer` takes care of error handling for the entire function. If this steps gives an error, it will log or raise it. What I can do to improve the error handling is to separate the API call and the get operation.